Code review #3

giovannivelludo opened the pull request:
Class `BroadcastReceiver` is part of the replicated dictionary made for the first version of geocalendar, it interprets every received SMS and acts according to its content.

If you have any trouble understanding how backslashes are added and removed you can take a look at https://github.com/EIS0/network-dictionary/blob/code-review/giovannivelludo-readded-files/networklibrary/src/main/java/com/eis/smsnetwork/SMSNetDictionary.java. I didn't include that class in this review because there would have been too many lines of code.

Calls to `Log` were left because they can be removed automatically by ProGuard when building a release of the application using this library.

giobbarca commented on networklibrary/src/main/java/com/eis/smsnetwork/broadcast/BroadcastReceiver.java:
```diff
@@ -0,0 +1,182 @@
+        try {
+            request = RequestType.get(fields[0]);
+        } catch (ArrayIndexOutOfBoundsException | NullPointerException e) {
+            return;
```
I think it's a bad style idea to return from a catch block without any sort of exception managing or calling class alerting.

giovannivelludo replied:
The launch of all exceptions caught in this method is caused by the SMS we received not being formatted properly, so there's no need to alert any class, we just received an SMS that seemed to be correct when looking at the first fields, but then turned out to be garbage. Do you think I should call `Log` whenever we receive any such message?

giobbarca replied:
I think could be a good idea to call `Log` during dev stages and then, as you said in the PR description, use ProGuard to delete them in release stage and I think that because at least you're aware of what's happening in case some exception is thrown (if it's thrown in an "unexpected"  case).

giovannivelludo replied:
Ok

giobbarca commented on networklibrary/src/main/java/com/eis/smsnetwork/broadcast/BroadcastReceiver.java:
```diff
@@ -0,0 +1,182 @@
+                invitedPeers.remove(sender);
+
+                // Sending to the invited peer my subscribers list
+                StringBuilder mySubscribers = new StringBuilder(RequestType.AddPeer.asString() +
```
Although is not part of this review, I would have overridden `toString()` method instead of using `asString()` method, supposing that the purpose of this method is the same.

giovannivelludo replied:
`toString()` would return the `String` "AddPeer", and I think that's the default behavior of this method in `enums`, however we want to get the single character associated to a `RequestType` instead. I guess the method `asString()` was created because somebody might want to call `toString()` and have it behave as it does by default.
Anyway I didn't write method `asString()` in class `RequestType`.

![Screenshot from 2020-03-03 16-39-36](https://user-images.githubusercontent.com/56405609/75792033-b3591680-5d6d-11ea-9d3c-73d8f01278a4.png)

giobbarca replied:
Thank you for the explanation, it wasn't very clear to me! Nothing to say then.

giobbarca commented on networklibrary/src/main/java/com/eis/smsnetwork/broadcast/BroadcastReceiver.java:
```diff
@@ -0,0 +1,182 @@
+                for (SMSPeer peerToAdd : subscribers.getSubscribers())
+                    mySubscribers.append(peerToAdd).append(FIELD_SEPARATOR);
+                SMSMessage mySubscribersMessage = new SMSMessage(
+                        sender, mySubscribers.deleteCharAt(mySubscribers.length() - 1).toString());
```
Although it's evident from the code, I would have explained with a comment why do you remove the last char of the `mySubscribers` string to make this code more "readable"

giovannivelludo replied:
Ok, I'll do it

giovannivelludo commented:
> Almost all the code is self-explanatory, but in very few cases I would have written more comments to explain an hard-understandable code.

Could you point out where?

> I don't have nothing to say about the testing class, which is composed by quite simple and repetitive methods that are readable and correct overall.
Probably, even if they are self-explanatory, I would have specified what every method tests.

I thought the name of each test method already explained what it does. Do you think I should add a comment explaining what they do?

giobbarca commented:
> > Almost all the code is self-explanatory, but in very few cases I would have written more comments to explain an hard-understandable code.
> 
> Could you point out where?

The only doubt I had about this was resolved with your last commit. I thought come cases in the switch needed more explanation but it's evincible from the code itself and from the method JavaDoc. Good job.

> > I don't have nothing to say about the testing class, which is composed by quite simple and repetitive methods that are readable and correct overall.
> > Probably, even if they are self-explanatory, I would have specified what every method tests.
> 
> I thought the name of each test method already explained what it does. Do you think I should add a comment explaining what they do?

I would have provided "human-readable" examples of what each test can accept or reject.
Something like:
`// Checks if message is valid'`
`// "aaabbbccc" is valid`
`// "aaabbbccc/" is not valid`

giovannivelludo commented:
@giobbarca is everything fine now?

giobbarca commented:
Now I think it's perfect and I really like how you used {@code} in the `onMessageReceived` method's JavaDocs (I didn't know it could be used also in that way, thank you).

giovannivelludo commented:
I just pushed another commit because I wanted to improve coverage of tests in the branch I'm currently working on, and while doing that I noticed that some `catch` blocks were useless, so I removed them in here too (and added the tests I added in the other branch to improve coverage). Since this last commit modifies around 500 lines of code and this pull request already contained 700 lines of code when I sent it to you, I don't think you need to review it, as it would be too much work.
